src/Http/WebUtilities/src/FormPipeReader.cs

```diff
@@ -0,0 +1,342 @@
+                    span = span.Slice(ampersand + andDelimiter.Length);
+                }
+
+                var decodedKey = GetDecodedString(key);
+                var decodedValue = GetDecodedString(value);
+
+                AppendAndVerify(ref accumulator, decodedKey, decodedValue);
+
+                // Cover case where we don't have an ampersand at the end.
+                consumed += key.Length + value.Length + (ampersand == -1 ? equalsDelimiter.Length : equalsDelimiter.Length + andDelimiter.Length);
+            }
+        }
+
+        // For multi-segment parsing of a read only sequence
+        private void ParseValuesSlow(
+            ref ReadOnlySequence<byte> buffer,
+            ref KeyValueAccumulator accumulator,
+            bool isFinalBlock)
+        {
+            var sequenceReader = new SequenceReader<byte>(buffer);
+            var consumed = sequenceReader.Position;
+            var equalsDelimiter = GetEqualsForEncoding();
+            var andDelimiter = GetAndForEncoding();
+
+            while (!sequenceReader.End)
+            {
+                // TODO seems there is a bug with TryReadTo (advancePastDelimiter: true). It isn't advancing past the delimiter on second read.
+                if (!sequenceReader.TryReadTo(out ReadOnlySequence<byte> key, equalsDelimiter, advancePastDelimiter: false) ||
+                    !sequenceReader.IsNext(equalsDelimiter, true))
+                {
+                    break;
+                }
+
+                if (!sequenceReader.TryReadTo(out ReadOnlySequence<byte> value, andDelimiter, false) ||
+                    !sequenceReader.IsNext(andDelimiter, true))
+                {
+                    if (!isFinalBlock)
+                    {
+                        break;
+                    }
+
+                    value = buffer.Slice(sequenceReader.Position);
+
+                    sequenceReader.Advance(value.Length);
+                }
+
+                // Need to call ToArray if the key/value spans multiple segments 
+                var decodedKey = GetDecodedStringFromReadOnlySequence(key);
+                var decodedValue = GetDecodedStringFromReadOnlySequence(value);
+
+                AppendAndVerify(ref accumulator, decodedKey, decodedValue);
+
+                consumed = sequenceReader.Position;
+            }
+
+            buffer = buffer.Slice(consumed);
+        }
+
+        private string GetDecodedStringFromReadOnlySequence(ReadOnlySequence<byte> ros)
+        {
+            if (ros.IsSingleSegment)
+            {
+                return GetDecodedString(ros.First.Span);
+            }
+
+            if (ros.Length < StackAllocThreshold)
+            {
+                Span<byte> buffer = stackalloc byte[(int)ros.Length];
+                ros.CopyTo(buffer);
+                return GetDecodedString(buffer);
+            }
+            else
+            {
+                var byteArray = ArrayPool<byte>.Shared.Rent((int)ros.Length);
```

Comment:
Should this be returned in a finally block or is that not necessary?

Reply:
In this can, I don't think we need to, but it's generally a good practice. I'll go ahead and add it.